python/lsst/pipe/tasks/calibrateImage.py

```diff
@@ -631,34 +640,71 @@ def run(self, *, exposures, id_generator=None, result=None):
+        # masquerade as the real thing.
+        have_fit_psf = False
+        have_fit_astrometry = False
+        have_fit_photometry = False
+        try:
+            result.psf_stars_footprints, result.background, _ = self._compute_psf(result.exposure,
+                                                                                  id_generator)
+            have_fit_psf = True
+            summary_stat_catalog = result.psf_stars_footprints
+            self._measure_aperture_correction(result.exposure, result.psf_stars_footprints)
+            result.psf_stars = result.psf_stars_footprints.asAstropy()
+
+            result.stars_footprints = self._find_stars(result.exposure, result.background, id_generator)
+            self._match_psf_stars(result.psf_stars_footprints, result.stars_footprints)
+            summary_stat_catalog = result.stars_footprints
+            result.stars = result.stars_footprints.asAstropy()
+            self.metadata["star_count"] = np.sum(~result.stars["sky_source"])
+
+            astrometry_matches, astrometry_meta = self._fit_astrometry(
+                result.exposure, result.stars_footprints
+            )
+            have_fit_astrometry = True
+            self.metadata["astrometry_matches_count"] = len(astrometry_matches)
+            if "astrometry_matches" in self.config.optional_outputs:
+                result.astrometry_matches = lsst.meas.astrom.denormalizeMatches(astrometry_matches,
+                                                                                astrometry_meta)
+
+            result.stars_footprints, photometry_matches, \
+                photometry_meta, photo_calib = self._fit_photometry(result.exposure, result.stars_footprints)
+            have_fit_photometry = True
+            self.metadata["photometry_matches_count"] = len(photometry_matches)
+            # fit_photometry returns a new catalog, so we need a new astropy table view.
+            result.stars = result.stars_footprints.asAstropy()
+            # summary stats don't make use of the calibrated fluxes, but we
+            # might as well use the best catalog we've got in case that
+            # changes, and help the old one get garbage-collected.
+            summary_stat_catalog = result.stars_footprints
+            if "photometry_matches" in self.config.optional_outputs:
+                result.photometry_matches = lsst.meas.astrom.denormalizeMatches(photometry_matches,
+                                                                                photometry_meta)
+        except pipeBase.AlgorithmError:
+            if not have_fit_psf:
+                result.exposure.setPsf(None)
+            if not have_fit_astrometry:
+                result.exposure.setWcs(None)
+            if not have_fit_photometry:
+                result.exposure.setPhotoCalib(None)
+            # Summary stat calculations can handle missing components gracefully,
```

Comment:
This and the other long comments are useful for understanding the logic, but I'm definitely worried that this commit requires a couple of paragraphs of comments to explain why it's implemented this way.